File: 카카오/2020카카오여름인턴십_4.cpp
#include<bits/stdc++.h>

using namespace std;

struct point {
    int x, y;
    int ox, oy;
    int  val;
};

int dx[] = { -1, 0, 1, 0 };
int dy[] = { 0, 1, 0, -1 };

int visited[27][27];

int solution(vector<vector<int>> board) {
    int answer = INT_MAX;
    int size = board.size();
    queue<point> q;

    q.push({ 0, 0, 0, 0, 0 });
    visited[0][0] = 0;

    while (!q.empty()) {
        point tmp = q.front();
        q.pop();

        if (tmp.x == size - 1 && tmp.y == size - 1) {
            answer = min(answer, tmp.val);
            continue;
        }

        for (int i = 0; i < 4; i++) {
            int nx = tmp.x + dx[i];
            int ny = tmp.y + dy[i];
            int nval = tmp.val + 100;

            if (tmp.ox != nx && tmp.oy != ny) {
                nval += 500;
            }

            if (nx < 0 || nx >= size || ny < 0 || ny >= size || board[nx][ny] == 1) {
                continue;
            }

            if (visited[nx][ny] != 0 && visited[nx][ny] < nval) {
                continue;
            }

            q.push({ nx, ny, tmp.x, tmp.y, nval });
            visited[nx][ny] = nval;
        }
    }
    return answer;
}

int main() {
    cout << solution({ {0,0,1, 0}, {0, 0,0,0}, {0,1, 0,1}, {1,0,0,0} });
}

File: 카카오/2020카카오여름인턴십_4.java
import java.lang.reflect.Array;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

class Solution {
    public int solution(int[][] board) {
        int size = board.length;
        int[] dx = {-1, 0, 1, 0};
        int[] dy = {0, -1, 0, 1};
        int[][] visit = new int[size][size];

        for (int[] v : visit) {
            Arrays.fill(v, Integer.MAX_VALUE);
        }

        Queue<Car> queue = new LinkedList<>();
        queue.add(new Car(0, 0, 0, 0, 0));
        visit[0][0] = 0;

        int answer = Integer.MAX_VALUE;
        while (!queue.isEmpty()) {
            Car car = queue.poll();
            if (car.x == size - 1 && car.y == size - 1) {
                answer = Math.min(answer, car.value);
                continue;
            }

            for (int i = 0; i < 4; i++) {
                int nx = car.x + dx[i];
                int ny = car.y + dy[i];
                int nValue = car.value + 100;

                if (nx < 0 || nx >= size || ny < 0 || ny >= size || board[nx][ny] == 1 || visit[nx][ny] < nValue) {
                    continue;
                }

                if (car.ox != nx && car.oy != ny) {
                    nValue += 500;
                }

                queue.add(new Car(nx, ny, car.x, car.y, nValue));
                visit[nx][ny] = nValue;
            }
        }
        return answer;
    }

    private static class Car {
        int x, ox;
        int y, oy;
        int value;

        public Car(int x, int y, int ox, int oy, int value) {
            this.x = x;
            this.y = y;
            this.ox = ox;
            this.oy = oy;
            this.value = value;
        }
    }

}
